### PR TITLE
Add FileUtility, DirectoryUtility.

### PR DESCRIPTION
**Pull request by ddunkin**
Are these generally useful? They can only exist in net46 and netstandard2.0. `Find` may be problematic where symlink cycles are possible. `TryDelete` is used a lot in Amber.

### CODE REVIEW

File: src/Faithlife.Utility/DirectoryUtility.cs [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+		private static IEnumerable<string> DoFindFiles(string path, string searchPattern)
+		{
+			IEnumerable<string> files = Directory.GetFiles(path, searchPattern, SearchOption.TopDirectoryOnly);
+#if NET46
+			files = files.Where(str =>
+			{
+				try
+				{
+					new FileIOPermission(FileIOPermissionAccess.PathDiscovery, str).Demand();
```
Comment (bgrainger) [inline src/Faithlife.Utility/DirectoryUtility.cs]: We should remove CAS.

**Comment by bgrainger** [comment]
What's the blocker for `netstandard1.4`? A quick peek at https://docs.microsoft.com/en-us/dotnet/api/?term=File.Delete&view=netstandard-1.4 indicates that at least some APIs should be available there.

**Comment by ddunkin** [comment]
My mistake, `File` is added in netstandard1.3. I'll fix this up.

File: src/Faithlife.Utility/DirectoryUtility.cs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+		/// <summary>
+		/// Recursively finds all files matching <paramref name="searchPatterns"/> in the folders specified by
+		/// <paramref name="directoryPaths"/> and their subfolders specified by <paramref name="getDirectories"/>.
+		/// </summary>
+		/// <param name="directoryPaths">The directory paths to search.</param>
+		/// <param name="searchPatterns">The search patterns to match against the names of files.</param>
+		/// <param name="getDirectories">A function which takes a path to a directory and returns a list of subdirectories of that path.</param>
+		/// <param name="cancellationToken">The cancellation token.</param>
+		/// <returns>A sequence of FileInfo instances.</returns>
+		public static IEnumerable<FileInfo> FindFileInfos(IEnumerable<string> directoryPaths, IEnumerable<string> searchPatterns, Func<string, string[]> getDirectories, CancellationToken cancellationToken)
```
Comment (ejball) [inline src/Faithlife.Utility/DirectoryUtility.cs]: Is this intentionally public? How is it meant to be used?

File: src/Faithlife.Utility/DirectoryUtility.cs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+		/// Recursively finds all files matching <paramref name="searchPatterns"/> in the folders specified by
+		/// <paramref name="directoryPaths"/> and their subfolders.
+		/// </summary>
+		/// <param name="directoryPaths">The directory paths to search.</param>
+		/// <param name="searchPatterns">The search patterns to match against the names of files.</param>
+		/// <param name="cancellationToken">The cancellation token.</param>
+		/// <returns>A sequence of fully-qualified file paths.</returns>
+		public static IEnumerable<string> FindFiles(IEnumerable<string> directoryPaths, IEnumerable<string> searchPatterns, CancellationToken cancellationToken)
+		{
+			return DoFindFiles(directoryPaths, searchPatterns, DoFindFiles, Directory.GetDirectories, cancellationToken);
```
Comment (ejball) [inline src/Faithlife.Utility/DirectoryUtility.cs]: Shouldn't we use `Directory.EnumerateDirectories`? And allow cancellation as we enumerate?

File: src/Faithlife.Utility/DirectoryUtility.cs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+				foreach (string childDirectoryPath in directories.Reverse())
+				{
+					// TODO: check for cycles
+					pathsStack.Push(childDirectoryPath);
+				}
+			}
+		}
+
+		private static IEnumerable<string> DoFindFiles(string path, string searchPattern) => Directory.GetFiles(path, searchPattern, SearchOption.TopDirectoryOnly);
```
Comment (ejball) [inline src/Faithlife.Utility/DirectoryUtility.cs]: Shouldn't we use `Directory.EnumerateFiles`?

File: src/Faithlife.Utility/DirectoryUtility.cs (hunk 2 above)
Comment (ddunkin) [reply src/Faithlife.Utility/DirectoryUtility.cs]: It's used in several places in internal resource tools. Not general purpose enough? Same with `FindFiles`.

Comment (ddunkin) [reply src/Faithlife.Utility/DirectoryUtility.cs]: After further review, these are confusing, rarely used, and predate `EnumerateFiles`. I'm go to remove them.

**Pull request merged**